################################################################################
#
# 1_data_preproc.R
#
# Pre-processing of 1000 Genomes and Gourraud2014 datasets for later comparison
# of genotypes and allele frequencies
#
# Oct/2014
#
################################################################################
#----------------------------------------------------------
# 1. Organizing PAG2014 dataset

# data directory
data_dir <- "data/"

# PAG2014 data
pg <- read.table(paste0(data_dir, "mhc.tab"), sep = "\t",
		   header = TRUE, stringsAsFactors = FALSE, row.names = 1)

# Discard CHD individuals, which are not in 1000G Phase I
pg <- pg[- grep("Chinese from Denver-Colorado, USA$", pg$Population), ]

# erase "XX" in the allele name "14XX"
pg[pg=="14XX"] <- "14"

# Remove _1 and _2 from subject names
pg$Subject <- gsub("_.", "", pg$Subject)

# get list of all 1000G individuals
if (! file.exists(paste0(data_dir,
                         "phase1_integrated_calls.20101123.ALL.panel"))) {
    download.file("ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20110521/phase1_integrated_calls.20101123.ALL.panel",
                  destfile = paste0(data_dir,
                                "phase1_integrated_calls.20101123.ALL.panel"))
}

indkg <- read.table(paste0(data_dir,
                            "phase1_integrated_calls.20101123.ALL.panel"),
                     header = F, fill = T, as.is=T, col.names = c("IND", "POP",
                                            "CONT", "PLAT1", "PLAT2", "PLAT3"))

indkg <- indkg[,c("POP", "IND")]

# remove individuals not in 1000G
pg <- pg[pg$Subject %in% indkg$IND, ]

# Load consensus sequences for HLA alleles at Pierre's data resolution
# generated by Vitor's script pierre2.R
load(paste0(data_dir, "cons_seq_pag.rda"))

# Organize consensus sequences in list
cons_seq_pierre <- list()
loci <- c("A" , "B" , "C" , "DQB1" , "DRB1")
for(l in loci){
    cons_seq_pierre[[l]] <- get(paste0("cons_seq_pierre_",l))
}

# Pierre object for comparison with 1000G:
pag2014 <- list()
loci <- c("A" , "B" , "C" , "DQB1" , "DRB1")
for(l in loci){
    pag2014[[l]] <- sapply(pier[,paste0("HLA.", l, ".", 1:2)], function(x)
                          cons_seq_pierre[[l]][x, ], simplify = "array")
    rownames(pag2014[[l]]) <- pier$Subject
    pag2014[[l]] <- pag2014[[l]][order(rownames(pag2014[[l]])),,]
    # remove ambiguities with /* or */
    pag2014[[l]] <- gsub("/\\*", "", pag2014[[l]])
    pag2014[[l]] <- gsub("\\*/", "", pag2014[[l]])
}

#----------------------------------------------------------
# 2. Organizing 1000G dataset

# Coordinates of ARS exons in hg19.
# Note that some HLA loci are on the - strand
exars_hg19 <- list()
exars_hg19$A <- c(29910534:29910803,29911045:29911320)
exars_hg19$B <- c(31324734:31324465,31324219:31323944)
exars_hg19$C <- c(31239645:31239376,31239125:31238850)
exars_hg19$DQB1 <- c(32632844:32632575)
exars_hg19$DRB1 <- c(32552155:32551886)

# write file with the names of all individuals in both Pierre's and 1000G data
# write.table(pier$Subject, file = paste0(data_dir,
# "samples_overlap_kg_pag2014.txt"), quote = F, row.names = F, col.names = F)

# Then use vcftools to extract those individuals and positions in ARS exons from
# chr6 vcf.gz file in pedmap format as in:
# vcftools 
# --gzvcf ~/hla_tools/1kg/data/phase1_chr/ALL.chr6.phase1_release_v3.20101123.snps_indels_svs.genotypes.vcf.gz
# --keep ~/hla_tools/1kg/data/samples_overlap_kg_pag2014.txt
# --out ~/hla_tools/1kg/data/a_ars_pag2014ind
# --plink --remove-indels --bed ~/hla_tools/1kg/data/a_ex23.bed

#...and read those files here:

# Reading pedmap files generated by vcftools site and sample filtering
positions <- list()
alleles  <- list()

for (l in loci){
    positions[[l]] <- read.table(paste0(data_dir, tolower(l),
                                        "_ars_pag2014ind.map"))[,4]
    alleles[[l]] <- as.matrix(read.table(paste0(data_dir, tolower(l),
                                                "_ars_pag2014ind.ped"),
                                  colClasses = "character"))
}

#Creating kg object for comparison with PAG2014 data
kg <- list()
for (l in loci){
    kg[[l]] <- array(dim = c(nrow(pag2014[[l]]), ncol(pag2014[[l]]), 2) ,
                     dimnames = list(individual = alleles[[l]][,1],
                                     position = 1:ncol( pag2014[[l]]),
                                     allele = 1:2))
    kg[[l]][, match( positions[[l]], exars_hg19[[l]] ), 1] <-
        alleles[[l]] [, seq(7, ncol( alleles[[l]]), 2) ]
    kg[[l]][, match( positions[[l]] , exars_hg19[[l]] ), 2] <-
        alleles[[l]] [, seq(8, ncol( alleles[[l]]), 2) ]
}

# HLA-B -C -DQB1 and -DRB1 are on the minus strand. To allow comparison with
# the Sanger data, convert nucleotides in the 1000G, which by convention are
# reported on the + strand to the - strand

kg[c("B", "C", "DQB1", "DRB1")] <- lapply(kg[c("B", "C", "DQB1", "DRB1")],
    function(x){ chartr(new="ATCG", old="TAGC", x)-> x }
)

#-------------------------------------------------------------------------------
# 3. Compare polymorphic positions in both datasets

polim.kg <- list()
polim.pag <- list()

for(l in loci){
    # Polymorphic positions in KG
    polim.kg[[l]] <- which(apply(rbind(kg[[l]][,,1], kg[[l]][,,2] ), 2,
                                 function(x) length(unique(x))) > 1)
    # Polymorphic positions in PAG2014
    polim.pag[[l]] <- which(apply(rbind(pag2014[[l]][,,1], pag2014[[l]][,,2]),
                                  2, function(x) length(unique(x[x!="*"]))) > 1)
}

###################################################################################
# Save Workspace
save.image("1_data_preproc.rda")
